### PR TITLE
Autofire that isn't objectively terrible. We truly live in a society.

### PR DESCRIPTION
**Pull request by Kmc2000**
## About The Pull Request

Completely tears out our old autofire code with extreme prejudice.

## Why It's Good For The Game

It's like autofire always used to be, except it actually works now

## Changelog
:cl: Kmc2000 + Gun Jesus
add: Fully automatic firing has been fixed, and reduced to only a small set of guns.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/autofire.dm [hunk 1]
```diff
@@ -1,70 +1,130 @@
...
+					autofire_component.RemoveComponent()
+					qdel(autofire_component)
+					return ..()
+		if(NAMEOF(src, fire_rate))
+			autofire_component?.default_fire_delay = (10 / var_value)
+			return ..()
+		else
+			return ..()
+
+//Place any guns that you want to be fully automatic here (for record-keeping and so NSV can avoid conflicts please and thank.)
```
Comment (PowerfulBacon) [inline code/modules/projectiles/autofire.dm:43]: what if you made this on beestation so there wouldn't be conflicts 🥺

File: code/modules/projectiles/autofire.dm [hunk 2]
```diff
@@ -1,70 +1,130 @@
...
+	. = ..()
+	switch(var_name)
+		if(NAMEOF(src, full_auto))
+			canMouseDown = var_value //Necessary to fully apply the component.
+			if(var_value) //Admin editing the autofire to true on a gun, let's help them out.
+				if(!autofire_component)
+					autofire_component = AddComponent(/datum/component/full_auto)
+					if(fire_rate)
+						autofire_component.default_fire_delay = (10 / fire_rate)
+					return ..()
```
Comment (PowerfulBacon) [inline code/modules/projectiles/autofire.dm:31]: You run ..() twice

File: code/modules/projectiles/autofire.dm (hunk 1 above)
Comment (Kmc2000) [reply code/modules/projectiles/autofire.dm:43]: stop stealing my ideas

File: code/modules/projectiles/autofire.dm (hunk 2 above)
Comment (Kmc2000) [reply code/modules/projectiles/autofire.dm:31]: True, none of this is explicitly necessary and I just did it for clarity

**Comment by Jalleo** [comment]
Not reviewing tonight but I want to see this test merged for at least one full weekend to see how many issues crop up. We should also add a full on notification for which guns have it so people can report bugs with them.

File: code/modules/projectiles/autofire.dm [hunk 3]
```diff
@@ -1,70 +1,139 @@
...
-		if(clumsy_check)//If theyre a clown, they can't autofire.
-			if (HAS_TRAIT(user, TRAIT_CLUMSY))
-				return FALSE
+*/
 
-	if(!can_shoot()) //Just because you can pull the trigger doesn't mean it can shoot. (*CLICK*)
+#define COMSIG_AUTOFIRE_END "stop_autofiring"
+
+/obj/item/gun
+	var/full_auto = FALSE //Set this if your gun uses full auto. ONLY guns that go brr should use this. Not pistols!
```
[the remark this answers is not in the document]
Comment (Kmc2000) [reply code/modules/projectiles/autofire.dm:18]: That old automatic var is on a lot of random guns, and I'm told it doesn't actually refer to guns that should have full auto

**Pull request merged**